db/src/excision.rs

```diff
@@ -127,3 +133,108 @@ pub(crate) fn excisions<'schema>(partition_map: &'schema PartitionMap, schema: &
         Ok(Some(excisions))
     }
 }
+
+pub(crate) fn enqueue_pending_excisions(conn: &rusqlite::Connection, schema: &Schema, tx_id: Entid, excisions: ExcisionMap) -> Result<()> {
+    // excisions.into_iter().map(|(entid, excision)| enqueue_pending_excision(self, entid, excision)).collect().and(Ok(()))
+
+    // if !excisions.is_none() {
+    //     bail!(DbError::NotYetImplemented(format!("Excision not yet implemented: {:?}", excisions)));
+    // }
+
+    let mut stmt1: rusqlite::Statement = conn.prepare("INSERT INTO excisions VALUES (?, ?, ?, ?)")?;
+    let mut stmt2: rusqlite::Statement = conn.prepare("INSERT INTO excision_attrs VALUES (?, ?)")?;
+
+    for (entid, excision) in excisions {
+        stmt1.execute(&[&entid, &excision.target, &excision.before_tx, &excision.before_tx.unwrap_or(tx_id)])?; // XXX
+        if let Some(attrs) = excision.attrs {
+            // println!("attrs {:?}", attrs);
+            for attr in attrs {
+                stmt2.execute(&[&entid, &attr])?;
+            }
+        }
+    }
+
+    // TODO: filter by attrs.
+    let mut stmt: rusqlite::Statement = conn.prepare(format!("WITH ids AS (SELECT d.rowid FROM datoms AS d, excisions AS e WHERE e.status > 0 AND (e.target IS d.e OR (e.target IS d.v AND d.a IS NOT {}))) DELETE FROM datoms WHERE rowid IN ids", entids::DB_EXCISE).as_ref())?;
```

Comment:
What's the reasoning behind splitting excision into two steps? e.g. first delete stuff from datoms, then on a separate call from transactions. It seems backwards to me (but perhaps understanding the reason for the split will help here) - e.g. we first clean out a materialization of our transactions (datoms), and only then the "source of truth". I'd expect a two-step operation to start with the source, and finish at the edges?

Reply:
This is a choice, suggested by Richard in https://github.com/mozilla/mentat/issues/21#issuecomment-372730559.

Excision is very expensive, so we want to be able to amortize the cost over time.  The initial implementation won't do much toward amortizing, but I want to put the framework in place for it.  The reason to start at the materialization (`datoms`, the edge) is that immediately after an excision the world presented to queries should change.  You could imagine an alternate implementation where the expensive writes to the `datoms` table are themselves queued, and instead the query engine grows knowledge of "excised entities" and filters them from query results as appropriate.  You could even do this with a SQLite view: if you had an index that made it fast to look up whether a particular entity was pending excision, like
```sql
CREATE INDEX idx_excisions_pending ON excisions (target) WHERE status IS NOT 0
```
then you could
```sql
CREATE VIEW datoms AS
SELECT * FROM datoms_before_excisions
WHERE NOT EXISTS
 (SELECT 1 FROM excisions
  WHERE
  excisions.status IS NOT 0 AND
  ((datoms.e IS excisions.target) OR (datoms.v IS excisions.target AND datoms.value_type_tag = 13))
```

That might be a good idea, since most of the time the `idx_excisions_pending` partial index will be empty and the view will be as performant (I think) as querying the underlying datoms directly.

However, the _really_ expensive part is rewriting all of the transaction log (and, later, rewriting the shared transaction log on a remote service).  We want to be able to do that a little at a time; hence the machinery.  So, really, we're splitting excision into `datoms` and `transactions`, and then splitting `transactions` into lots of little re-writing steps that can proceed incrementally.

Only time will tell which of these approaches is best, and whether the machinery is worth the effort.